File: android/src/com/android/tools/idea/actions/annotations/InferSupportAnnotationsAction.java
/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.actions.annotations;

import com.android.SdkConstants;
import com.android.tools.idea.gradle.dsl.model.GradleBuildModel;
import com.android.tools.idea.gradle.dsl.model.dependencies.ArtifactDependencyModel;
import com.android.tools.idea.gradle.dsl.model.dependencies.DependenciesModel;
import com.android.tools.idea.gradle.project.GradleProjectImporter;
import com.android.tools.idea.gradle.project.GradleSyncListener;
import com.android.tools.idea.gradle.util.Projects;
import com.android.tools.idea.model.AndroidModuleInfo;
import com.android.tools.idea.templates.RepositoryUrlManager;
import com.intellij.analysis.AnalysisScope;
import com.intellij.analysis.BaseAnalysisAction;
import com.intellij.analysis.BaseAnalysisActionDialog;
import com.intellij.codeInsight.FileModificationService;
import com.intellij.history.LocalHistory;
import com.intellij.history.LocalHistoryAction;
import com.intellij.ide.scratch.ScratchFileService;
import com.intellij.ide.scratch.ScratchRootType;
import com.intellij.lang.StdLanguages;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.Presentation;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.application.Result;
import com.intellij.openapi.command.WriteCommandAction;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.fileEditor.FileEditorManager;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.module.ModuleUtilCore;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.project.ProjectUtil;
import com.intellij.openapi.roots.ModuleRootModificationUtil;
import com.intellij.openapi.ui.Messages;
import com.intellij.openapi.util.Computable;
import com.intellij.openapi.util.Factory;
import com.intellij.openapi.util.Ref;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.*;
import com.intellij.refactoring.RefactoringBundle;
import com.intellij.usageView.UsageInfo;
import com.intellij.usageView.UsageViewUtil;
import com.intellij.usages.*;
import com.intellij.util.Processor;
import com.intellij.util.SequentialModalProgressTask;
import com.intellij.util.SequentialTask;
import com.intellij.util.containers.ContainerUtil;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.*;

import static com.android.tools.idea.gradle.dsl.model.dependencies.CommonConfigurationNames.COMPILE;
import static com.intellij.openapi.util.text.StringUtil.isNotEmpty;
import static com.intellij.openapi.util.text.StringUtil.pluralize;

/**
 * Analyze support annotations
 */
public class InferSupportAnnotationsAction extends BaseAnalysisAction {
  /** Whether this feature is enabled or not during development */
  static final boolean ENABLED = Boolean.valueOf(System.getProperty("studio.infer.annotations"));

  /** Number of times we pass through the project files */
  static final int MAX_PASSES = 3;

  @NonNls private static final String INFER_SUPPORT_ANNOTATIONS = "Infer Support Annotations";
  private static final int MAX_ANNOTATIONS_WITHOUT_PREVIEW = 5;

  public InferSupportAnnotationsAction() {
    super("Infer Support Annotations", INFER_SUPPORT_ANNOTATIONS);
    if (!ENABLED) {
      getTemplatePresentation().setVisible(false);
    }
  }

  private static final String ADD_DEPENDENCY = "Add Support Dependency";
  private static final int MIN_SDK_WITH_NULLABLE = 19;

  @Override
  public void update(AnActionEvent event) {
    if (!ENABLED) {
      return;
    }
    super.update(event);
    Project project = event.getProject();
    if (project == null || !Projects.isBuildWithGradle(project)) {
      Presentation presentation = event.getPresentation();
      presentation.setEnabled(false);
    }
  }

  @Override
  protected void analyze(@NotNull Project project, @NotNull AnalysisScope scope) {
    if (!Projects.isBuildWithGradle(project)) {
      return;
    }
    int[] fileCount = new int[] {0};
    PsiDocumentManager.getInstance(project).commitAllDocuments();
    final UsageInfo[] usageInfos = findUsages(project, scope, fileCount[0]);
    if (usageInfos == null) return;

    Map<Module, PsiFile> modules = findModulesFromUsage(usageInfos);

    if (!checkModules(project, scope, modules)) {
      return;
    }

    if (usageInfos.length < MAX_ANNOTATIONS_WITHOUT_PREVIEW) {
      ApplicationManager.getApplication().invokeLater(applyRunnable(project, () -> usageInfos));
    }
    else {
      showUsageView(project, usageInfos, scope);
    }
  }

  private static Map<Module, PsiFile> findModulesFromUsage(UsageInfo[] infos) {
    // We need 1 file from each module that requires changes (the file may be overwritten below):
    final Map<Module, PsiFile> modules = new HashMap<>();

    for (UsageInfo info : infos) {
      final PsiElement element = info.getElement();
      assert element != null;
      Module module = ModuleUtilCore.findModuleForPsiElement(element);
      if (module == null) {
        continue;
      }
      PsiFile file = element.getContainingFile();
      modules.put(module, file);
    }
    return modules;
  }

  private static UsageInfo[] findUsages(@NotNull final Project project,
                                        @NotNull final AnalysisScope scope,
                                        final int fileCount) {
    final InferSupportAnnotations inferrer = new InferSupportAnnotations(false, project);
    final PsiManager psiManager = PsiManager.getInstance(project);
    final Runnable searchForUsages = () -> scope.accept(new PsiElementVisitor() {
      int myFileCount = 0;

      @Override
      public void visitFile(final PsiFile file) {
        myFileCount++;
        final VirtualFile virtualFile = file.getVirtualFile();
        final FileViewProvider viewProvider = psiManager.findViewProvider(virtualFile);
        final Document document = viewProvider == null ? null : viewProvider.getDocument();
        if (document == null || virtualFile.getFileType().isBinary()) return; //do not inspect binary files
        final ProgressIndicator progressIndicator = ProgressManager.getInstance().getProgressIndicator();
        if (progressIndicator != null) {
          progressIndicator.setText2(ProjectUtil.calcRelativeToProjectPath(virtualFile, project));
          progressIndicator.setFraction(((double)myFileCount) / (MAX_PASSES * fileCount));
        }
        if (file instanceof PsiJavaFile) {
          inferrer.collect(file);
        }
      }
    });

    /*
      Collect these files and visit repeatedly. Consider this
      scenario, where I visit files A, B, C in alphabetical order.
      Let's say a method in A unconditionally calls a method in B
      calls a method in C. In file C I discover that the method
      requires permission P. At this point it's too late for me to
      therefore conclude that the method in B also requires it. If I
      make a whole separate pass again, I could now add that
      constraint. But only after that second pass can I infer that
      the method in A also requires it. In short, I need to keep
      passing through all files until I make no more progress. It
      would be much more efficient to handle this with a global call
      graph such that as soon as I make an inference I can flow it
      backwards.
     */
    Runnable multipass = () -> {
      for (int i = 0; i < MAX_PASSES; i++) {
        searchForUsages.run();
      }
    };

    if (ApplicationManager.getApplication().isDispatchThread()) {
      if (!ProgressManager.getInstance().runProcessWithProgressSynchronously(multipass, INFER_SUPPORT_ANNOTATIONS, true, project)) {
        return null;
      }
    } else {
      multipass.run();
    }

    final List<UsageInfo> usages = new ArrayList<>();
    inferrer.collect(usages);
    return usages.toArray(new UsageInfo[usages.size()]);
  }

  // For Android we need to check SDK version and possibly update the gradle project file
  protected boolean checkModules(@NotNull final Project project,
                                 @NotNull final AnalysisScope scope,
                                 @NotNull Map<Module, PsiFile> modules) {
    final Set<Module> modulesWithoutAnnotations = new HashSet<>();
    final Set<Module> modulesWithLowVersion = new HashSet<>();
    for (Module module : modules.keySet()) {
      AndroidModuleInfo info = AndroidModuleInfo.get(module);
      if (info != null && info.getBuildSdkVersion() != null && info.getBuildSdkVersion().getFeatureLevel() <  MIN_SDK_WITH_NULLABLE) {
        modulesWithLowVersion.add(module);
      }
      GradleBuildModel buildModel = GradleBuildModel.get(module);
      if (buildModel == null) {
        Logger.getInstance(InferSupportAnnotationsAction.class).warn("Unable to find Gradle build model for module " + module.getModuleFilePath());
        continue;
      }
      boolean dependencyFound = false;
      DependenciesModel dependenciesModel = buildModel.dependencies();
      if (dependenciesModel != null) {
        for (ArtifactDependencyModel dependency : dependenciesModel.artifacts(COMPILE)) {
          String notation = dependency.compactNotation().value();
          if (notation.startsWith(SdkConstants.APPCOMPAT_LIB_ARTIFACT) ||
              notation.startsWith(SdkConstants.ANNOTATIONS_LIB_ARTIFACT)) {
            dependencyFound = true;
            break;
          }
        }
      }
      if (!dependencyFound) {
        modulesWithoutAnnotations.add(module);
      }
    }

    if (!modulesWithLowVersion.isEmpty()) {
      Messages.showErrorDialog(
        project,
        String.format("Infer Support Annotations requires the project sdk level be set to %1$d or greater.", MIN_SDK_WITH_NULLABLE),
        "Infer Support Annotations");
      return false;
    }
    if (modulesWithoutAnnotations.isEmpty()) {
      return true;
    }
    String moduleNames = StringUtil.join(modulesWithoutAnnotations, Module::getName, ", ");
    int count = modulesWithoutAnnotations.size();
    String message = String.format("The %1$s %2$s %3$sn't refer to the existing '%4$s' library with Android nullity annotations. \n\n" +
                                   "Would you like to add the %5$s now?",
                                   pluralize("module", count),
                                   moduleNames,
                                   count > 1 ? "do" : "does",
                                   RepositoryUrlManager.SUPPORT_ANNOTATIONS,
                                   pluralize("dependency", count));
    if (Messages.showOkCancelDialog(project, message, "Infer Nullity Annotations", Messages.getErrorIcon()) == Messages.OK) {
      final LocalHistoryAction action = LocalHistory.getInstance().startAction(ADD_DEPENDENCY);
      try {
        new WriteCommandAction(project, ADD_DEPENDENCY) {
          @Override
          protected void run(@NotNull final Result result) throws Throwable {
            RepositoryUrlManager manager = RepositoryUrlManager.get();
            String annotationsLibraryCoordinate = manager.getLibraryCoordinate(RepositoryUrlManager.SUPPORT_ANNOTATIONS);
            for (Module module : modulesWithoutAnnotations) {
              addDependency(module, annotationsLibraryCoordinate);
            }
            GradleProjectImporter.getInstance().requestProjectSync(project, false, new GradleSyncListener.Adapter() {
              @Override
              public void syncSucceeded(@NotNull Project project) {
                restartAnalysis(project, scope);
              }
            });
          }
        }.execute();
      }
      finally {
        action.finish();
      }
    }
    return false;
  }

  private static Runnable applyRunnable(final Project project, final Computable<UsageInfo[]> computable) {
    return () -> {
      final LocalHistoryAction action = LocalHistory.getInstance().startAction(INFER_SUPPORT_ANNOTATIONS);
      try {
        new WriteCommandAction(project, INFER_SUPPORT_ANNOTATIONS) {
          @Override
          protected void run(@NotNull Result result) throws Throwable {
            final UsageInfo[] infos = computable.compute();
            if (infos.length > 0) {

              final Set<PsiElement> elements = new LinkedHashSet<>();
              for (UsageInfo info : infos) {
                final PsiElement element = info.getElement();
                if (element != null) {
                  PsiFile containingFile = element.getContainingFile();
                  // Skip results in .class files; these are typically from extracted AAR files
                  VirtualFile virtualFile = containingFile.getVirtualFile();
                  if (virtualFile.getFileType().isBinary()) {
                    continue;
                  }

                  ContainerUtil.addIfNotNull(elements, containingFile);
                }
              }
              if (!FileModificationService.getInstance().preparePsiElementsForWrite(elements)) return;

              final SequentialModalProgressTask progressTask = new SequentialModalProgressTask(project, INFER_SUPPORT_ANNOTATIONS, false);
              progressTask.setMinIterationTime(200);
              progressTask.setTask(new AnnotateTask(project, progressTask, infos));
              ProgressManager.getInstance().run(progressTask);
            } else {
              InferSupportAnnotations.nothingFoundMessage(project);
            }
          }
        }.execute();
      }
      finally {
        action.finish();
      }
    };
  }

  private void restartAnalysis(final Project project, final AnalysisScope scope) {
    ApplicationManager.getApplication().invokeLater(() -> analyze(project, scope));
  }

  private static void showUsageView(@NotNull Project project, final UsageInfo[] usageInfos, @NotNull AnalysisScope scope) {
    final UsageTarget[] targets = UsageTarget.EMPTY_ARRAY;
    final Ref<Usage[]> convertUsagesRef = new Ref<>();
    if (!ProgressManager.getInstance().runProcessWithProgressSynchronously(() -> ApplicationManager.getApplication().runReadAction(() -> {
      convertUsagesRef.set(UsageInfo2UsageAdapter.convert(usageInfos));
    }), "Preprocess Usages", true, project)) return;

    if (convertUsagesRef.isNull()) return;
    final Usage[] usages = convertUsagesRef.get();

    final UsageViewPresentation presentation = new UsageViewPresentation();
    presentation.setTabText("Infer Nullity Preview");
    presentation.setShowReadOnlyStatusAsRed(true);
    presentation.setShowCancelButton(true);
    presentation.setUsagesString(RefactoringBundle.message("usageView.usagesText"));

    final UsageView usageView = UsageViewManager.getInstance(project).showUsages(targets, usages, presentation, rerunFactory(project, scope));

    final Runnable refactoringRunnable = applyRunnable(project, () -> {
      final Set<UsageInfo> infos = UsageViewUtil.getNotExcludedUsageInfos(usageView);
      return infos.toArray(new UsageInfo[infos.size()]);
    });

    String canNotMakeString = "Cannot perform operation.\nThere were changes in code after usages have been found.\nPlease perform operation search again.";

    usageView.addPerformOperationAction(refactoringRunnable, INFER_SUPPORT_ANNOTATIONS, canNotMakeString, INFER_SUPPORT_ANNOTATIONS, false);
  }

  @NotNull
  private static Factory<UsageSearcher> rerunFactory(@NotNull final Project project, @NotNull final AnalysisScope scope) {
    return () -> new UsageInfoSearcherAdapter() {
      @Override
      protected UsageInfo[] findUsages() {
        return InferSupportAnnotationsAction.findUsages(project, scope, scope.getFileCount());
      }

      @Override
      public void generate(@NotNull Processor<Usage> processor) {
        processUsages(processor, project);
      }
    };
  }

  private static void addDependency(@NotNull final Module module, @Nullable final String libraryCoordinate) {
    if (isNotEmpty(libraryCoordinate)) {
      ModuleRootModificationUtil.updateModel(module, model -> {
        GradleBuildModel buildModel = GradleBuildModel.get(module);
        if (buildModel != null) {
          buildModel.dependencies().addArtifact(COMPILE, libraryCoordinate);
        }
      });
    }
  }

  /* Android nullable annotations do not support annotations on local variables. */
  @Override
  protected JComponent getAdditionalActionSettings(Project project, BaseAnalysisActionDialog dialog) {
    if (!Projects.isBuildWithGradle(project)) {
      return super.getAdditionalActionSettings(project, dialog);
    }
    return null;
  }

  private static class AnnotateTask implements SequentialTask {
    private final Project myProject;
    private UsageInfo[] myInfos;
    private final SequentialModalProgressTask myTask;
    private int myCount = 0;
    private final int myTotal;

    public AnnotateTask(Project project, SequentialModalProgressTask progressTask, UsageInfo[] infos) {
      myProject = project;
      myInfos = infos;
      myTask = progressTask;
      myTotal = infos.length;
    }

    @Override
    public void prepare() {
    }

    @Override
    public boolean isDone() {
      return myCount > myTotal - 1;
    }

    @Override
    public boolean iteration() {
      final ProgressIndicator indicator = myTask.getIndicator();
      if (indicator != null) {
        indicator.setFraction(((double)myCount) / myTotal);
      }

      InferSupportAnnotations.apply(myProject, myInfos[myCount++]);

      boolean done = isDone();

      if (isDone()) {
        try {
          showReport();
        } catch (Throwable ignore) {
        }
      }
      return done;
    }

    @Override
    public void stop() {
    }

    public void showReport() {
      if (InferSupportAnnotations.CREATE_INFERENCE_REPORT) {
        String report = InferSupportAnnotations.generateReport(myInfos);
        String fileName = "Annotation Inference Report";
        ScratchFileService.Option option = ScratchFileService.Option.create_new_always;
        VirtualFile f = ScratchRootType.getInstance().createScratchFile(myProject, fileName, StdLanguages.TEXT, report, option);
        if (f != null) {
          FileEditorManager.getInstance(myProject).openFile(f, true);
        }
      }
    }
  }
}

File: android/src/com/android/tools/idea/actions/annotations/InferSupportAnnotationsAction.kt
/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.actions.annotations

import com.android.SdkConstants
import com.android.tools.idea.actions.annotations.InferSupportAnnotations.Companion.apply
import com.android.tools.idea.actions.annotations.InferSupportAnnotations.Companion.generateReport
import com.android.tools.idea.actions.annotations.InferSupportAnnotations.Companion.nothingFoundMessage
import com.android.tools.idea.gradle.dsl.api.GradleBuildModel
import com.android.tools.idea.gradle.dsl.api.dependencies.CommonConfigurationNames
import com.android.tools.idea.gradle.project.GradleProjectInfo
import com.android.tools.idea.gradle.repositories.RepositoryUrlManager.Companion.get
import com.android.tools.idea.gradle.util.GradleUtil
import com.android.tools.idea.model.AndroidModuleInfo
import com.android.tools.idea.projectsystem.GoogleMavenArtifactId
import com.android.tools.idea.projectsystem.ProjectSystemSyncManager
import com.android.tools.idea.projectsystem.getProjectSystem
import com.google.common.util.concurrent.FutureCallback
import com.google.common.util.concurrent.Futures
import com.google.common.util.concurrent.MoreExecutors
import com.intellij.analysis.AnalysisScope
import com.intellij.analysis.BaseAnalysisAction
import com.intellij.analysis.BaseAnalysisActionDialog
import com.intellij.codeInsight.FileModificationService
import com.intellij.facet.ProjectFacetManager
import com.intellij.history.LocalHistory
import com.intellij.ide.scratch.ScratchFileService
import com.intellij.ide.scratch.ScratchRootType
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.command.WriteCommandAction
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.fileTypes.PlainTextLanguage
import com.intellij.openapi.module.Module
import com.intellij.openapi.module.ModuleUtilCore
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.project.calcRelativeToProjectPath
import com.intellij.openapi.roots.ModuleRootModificationUtil
import com.intellij.openapi.ui.Messages
import com.intellij.openapi.util.Computable
import com.intellij.openapi.util.Factory
import com.intellij.openapi.util.Ref
import com.intellij.openapi.util.text.StringUtil
import com.intellij.psi.PsiDocumentManager
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiElementVisitor
import com.intellij.psi.PsiFile
import com.intellij.psi.PsiJavaFile
import com.intellij.psi.PsiManager
import com.intellij.refactoring.RefactoringBundle
import com.intellij.usageView.UsageInfo
import com.intellij.usageView.UsageViewUtil
import com.intellij.usages.Usage
import com.intellij.usages.UsageInfo2UsageAdapter
import com.intellij.usages.UsageInfoSearcherAdapter
import com.intellij.usages.UsageSearcher
import com.intellij.usages.UsageTarget
import com.intellij.usages.UsageViewManager
import com.intellij.usages.UsageViewPresentation
import com.intellij.util.Processor
import com.intellij.util.SequentialModalProgressTask
import com.intellij.util.SequentialTask
import org.jetbrains.android.facet.AndroidFacet
import org.jetbrains.android.facet.AndroidRootUtil
import org.jetbrains.android.refactoring.isAndroidx
import org.jetbrains.annotations.NonNls
import java.util.Locale
import javax.swing.JComponent

/** Analyze support annotations */
class InferSupportAnnotationsAction : BaseAnalysisAction("Infer Support Annotations", INFER_SUPPORT_ANNOTATIONS) {
  override fun update(event: AnActionEvent) {
    if (!ENABLED) {
      event.presentation.isEnabledAndVisible = false
      return
    }
    val project = event.project
    if (project == null || !ProjectFacetManager.getInstance(project).hasFacets(AndroidFacet.ID)) {
      // don't show this action in IDEA in non-android projects
      event.presentation.isEnabledAndVisible = false
      return
    }
    super.update(event)
    if (!GradleProjectInfo.getInstance(project).isBuildWithGradle) {
      val presentation = event.presentation
      presentation.isEnabled = false
    }
  }

  override fun analyze(project: Project, scope: AnalysisScope) {
    if (!GradleProjectInfo.getInstance(project).isBuildWithGradle) {
      return
    }
    val fileCount = intArrayOf(0)
    PsiDocumentManager.getInstance(project).commitAllDocuments()
    val usageInfos = findUsages(project, scope, fileCount[0]) ?: return
    val modules = findModulesFromUsage(usageInfos)
    if (!checkModules(project, scope, modules)) {
      return
    }
    if (usageInfos.size < MAX_ANNOTATIONS_WITHOUT_PREVIEW) {
      ApplicationManager.getApplication().invokeLater(applyRunnable(project) { usageInfos })
    } else {
      showUsageView(project, usageInfos, scope)
    }
  }

  // For Android we need to check SDK version and possibly update the gradle project file
  private fun checkModules(
    project: Project,
    scope: AnalysisScope,
    modules: Map<Module, PsiFile>
  ): Boolean {
    val modulesWithoutAnnotations: MutableSet<Module> = HashSet()
    val modulesWithLowVersion: MutableSet<Module> = HashSet()
    for (module in modules.keys) {
      val info = AndroidModuleInfo.getInstance(module)
      if (info != null && info.buildSdkVersion != null && info.buildSdkVersion!!.featureLevel < MIN_SDK_WITH_NULLABLE) {
        modulesWithLowVersion.add(module)
      }
      val buildModel = GradleBuildModel.get(module)
      if (buildModel == null) {
        Logger.getInstance(InferSupportAnnotationsAction::class.java)
          .warn("Unable to find Gradle build model for module " + AndroidRootUtil.getModuleDirPath(module))
        continue
      }
      var dependencyFound = false
      val dependenciesModel = buildModel.dependencies()
      val configurationName =
        GradleUtil.mapConfigurationName(CommonConfigurationNames.COMPILE, GradleUtil.getAndroidGradleModelVersionInUse(module), false)
      for (dependency in dependenciesModel.artifacts(configurationName)) {
        val notation = dependency.compactNotation()
        if (notation.startsWith(SdkConstants.APPCOMPAT_LIB_ARTIFACT) ||
          notation.startsWith(SdkConstants.ANDROIDX_APPCOMPAT_LIB_ARTIFACT) ||
          notation.startsWith(SdkConstants.SUPPORT_LIB_ARTIFACT) ||
          notation.startsWith(SdkConstants.ANDROIDX_SUPPORT_LIB_ARTIFACT) ||
          notation.startsWith(SdkConstants.ANDROIDX_ANNOTATIONS_ARTIFACT) ||
          notation.startsWith(SdkConstants.ANNOTATIONS_LIB_ARTIFACT)
        ) {
          dependencyFound = true
          break
        }
      }
      if (!dependencyFound) {
        modulesWithoutAnnotations.add(module)
      }
    }
    if (modulesWithLowVersion.isNotEmpty()) {
      Messages.showErrorDialog(
        project,
        String.format(
          Locale.US,
          "Infer Support Annotations requires the project sdk level be set to %1\$d or greater.",
          MIN_SDK_WITH_NULLABLE
        ),
        "Infer Support Annotations"
      )
      return false
    }
    if (modulesWithoutAnnotations.isEmpty()) {
      return true
    }
    val moduleNames = StringUtil.join(modulesWithoutAnnotations, { obj: Module -> obj.name }, ", ")
    val count = modulesWithoutAnnotations.size
    val message = String.format(
      """
      The %1${"$"}s %2${"$"}s %3${"$"}sn't refer to the existing '%4${"$"}s' library with Android nullity annotations.

      Would you like to add the %5${"$"}s now?
      """.trimIndent(),
      StringUtil.pluralize("module", count),
      moduleNames,
      if (count > 1) "do" else "does",
      GoogleMavenArtifactId.SUPPORT_ANNOTATIONS.mavenArtifactId,
      StringUtil.pluralize("dependency", count)
    )
    if (Messages.showOkCancelDialog(
        project,
        message,
        "Infer Nullity Annotations",
        "OK",
        "Cancel",
        Messages.getErrorIcon()
      ) == Messages.OK
    ) {
      val action = LocalHistory.getInstance().startAction(ADD_DEPENDENCY)
      try {
        WriteCommandAction.writeCommandAction(project).withName(ADD_DEPENDENCY).run<RuntimeException> {
          val manager = get()
          val annotation =
            if (project.isAndroidx()) GoogleMavenArtifactId.ANDROIDX_SUPPORT_ANNOTATIONS else GoogleMavenArtifactId.SUPPORT_ANNOTATIONS
          val annotationsLibraryCoordinate = manager.getArtifactStringCoordinate(annotation, true)
          if (annotationsLibraryCoordinate != null) {
            for (module in modulesWithoutAnnotations) {
              addDependency(module, annotationsLibraryCoordinate)
            }
          }
          syncAndRestartAnalysis(project, scope)
        }
      } finally {
        action.finish()
      }
    }
    return false
  }

  private fun syncAndRestartAnalysis(project: Project, scope: AnalysisScope) {
    assert(ApplicationManager.getApplication().isDispatchThread)
    val syncResult = project.getProjectSystem()
      .getSyncManager().syncProject(ProjectSystemSyncManager.SyncReason.PROJECT_MODIFIED)
    Futures.addCallback(
      syncResult,
      object : FutureCallback<ProjectSystemSyncManager.SyncResult> {
        override fun onSuccess(syncResult: ProjectSystemSyncManager.SyncResult?) {
          if (syncResult != null && syncResult.isSuccessful) {
            restartAnalysis(project, scope)
          }
        }

        override fun onFailure(t: Throwable?) {
          throw RuntimeException(t)
        }
      },
      MoreExecutors.directExecutor()
    )
  }

  private fun restartAnalysis(project: Project, scope: AnalysisScope) {
    ApplicationManager.getApplication().invokeLater { analyze(project, scope) }
  }

  /* Android nullable annotations do not support annotations on local variables. */
  override fun getAdditionalActionSettings(project: Project, dialog: BaseAnalysisActionDialog): JComponent? {
    return if (!GradleProjectInfo.getInstance(project).isBuildWithGradle) {
      super.getAdditionalActionSettings(project, dialog)
    } else null
  }

  private class AnnotateTask(
    private val myProject: Project,
    private val myTask: SequentialModalProgressTask,
    private val myInfos: Array<UsageInfo>
  ) : SequentialTask {
    private var myCount = 0
    private val myTotal: Int = myInfos.size

    override fun isDone(): Boolean {
      return myCount > myTotal - 1
    }

    override fun iteration(): Boolean {
      val indicator = myTask.indicator
      if (indicator != null) {
        indicator.fraction = myCount.toDouble() / myTotal
      }
      apply(myProject, myInfos[myCount++])
      val done = isDone
      if (isDone) {
        try {
          showReport()
        } catch (ignore: Throwable) {
        }
      }
      return done
    }

    fun showReport() {
      if (InferSupportAnnotations.CREATE_INFERENCE_REPORT) {
        val report = generateReport(myInfos)
        val fileName = "Annotation Inference Report"
        val option = ScratchFileService.Option.create_new_always
        val f = ScratchRootType.getInstance().createScratchFile(myProject, fileName, PlainTextLanguage.INSTANCE, report, option)
        if (f != null) {
          FileEditorManager.getInstance(myProject).openFile(f, true)
        }
      }
    }
  }

  companion object {
    /** Whether this feature is enabled or not during development */
    val ENABLED = java.lang.Boolean.parseBoolean("studio.infer.annotations")

    /** Number of times we pass through the project files */
    const val MAX_PASSES = 3
    private const val INFER_SUPPORT_ANNOTATIONS: @NonNls String = "Infer Support Annotations"
    private const val MAX_ANNOTATIONS_WITHOUT_PREVIEW = 5
    private const val ADD_DEPENDENCY = "Add Support Dependency"
    private const val MIN_SDK_WITH_NULLABLE = 19
    private fun findModulesFromUsage(infos: Array<UsageInfo>): Map<Module, PsiFile> {
      // We need 1 file from each module that requires changes (the file may be overwritten below):
      val modules: MutableMap<Module, PsiFile> = HashMap()
      for (info in infos) {
        val element = info.element ?: continue
        val module = ModuleUtilCore.findModuleForPsiElement(element) ?: continue
        val file = element.containingFile
        modules[module] = file
      }
      return modules
    }

    private fun findUsages(
      project: Project,
      scope: AnalysisScope,
      fileCount: Int
    ): Array<UsageInfo>? {
      val inferrer = InferSupportAnnotations(false, project)
      val psiManager = PsiManager.getInstance(project)
      val searchForUsages = Runnable {
        scope.accept(object : PsiElementVisitor() {
          var myFileCount = 0
          override fun visitFile(file: PsiFile) {
            myFileCount++
            val virtualFile = file.virtualFile
            val viewProvider = psiManager.findViewProvider(virtualFile)
            val document = viewProvider?.document
            if (document == null || virtualFile.fileType.isBinary) return // do not inspect binary files
            val progressIndicator = ProgressManager.getInstance().progressIndicator
            if (progressIndicator != null) {
              progressIndicator.text2 = calcRelativeToProjectPath(virtualFile, project)
              progressIndicator.fraction = myFileCount.toDouble() / (MAX_PASSES * fileCount)
            }
            if (file is PsiJavaFile) {
              inferrer.collect(file)
            }
          }
        })
      }

      /*
      Collect these files and visit repeatedly. Consider this
      scenario, where I visit files A, B, C in alphabetical order.
      Let's say a method in A unconditionally calls a method in B
      calls a method in C. In file C I discover that the method
      requires permission P. At this point it's too late for me to
      therefore conclude that the method in B also requires it. If I
      make a whole separate pass again, I could now add that
      constraint. But only after that second pass can I infer that
      the method in A also requires it. In short, I need to keep
      passing through all files until I make no more progress. It
      would be much more efficient to handle this with a global call
      graph such that as soon as I make an inference I can flow it
      backwards.
     */
      val multipass = Runnable {
        for (i in 0 until MAX_PASSES) {
          searchForUsages.run()
        }
      }
      if (ApplicationManager.getApplication().isDispatchThread) {
        if (!ProgressManager.getInstance().runProcessWithProgressSynchronously(multipass, INFER_SUPPORT_ANNOTATIONS, true, project)) {
          return null
        }
      } else {
        multipass.run()
      }
      val usages = ArrayList<UsageInfo>()
      inferrer.collect(usages, scope)
      return usages.toTypedArray()
    }

    private fun applyRunnable(project: Project, computable: Computable<Array<UsageInfo>>): Runnable {
      return Runnable {
        val action = LocalHistory.getInstance().startAction(INFER_SUPPORT_ANNOTATIONS)
        try {
          WriteCommandAction.writeCommandAction(project).withName(INFER_SUPPORT_ANNOTATIONS).run<RuntimeException> {
            val infos = computable.compute()
            if (infos.isNotEmpty()) {
              val elements: MutableSet<PsiElement> = LinkedHashSet()
              for (info in infos) {
                val element = info.element
                if (element != null) {
                  val containingFile = element.containingFile ?: continue
                  // Skip results in .class files; these are typically from extracted AAR files
                  val virtualFile = containingFile.virtualFile
                  if (virtualFile.fileType.isBinary) {
                    continue
                  }
                  elements.add(containingFile)
                }
              }
              if (!FileModificationService.getInstance().preparePsiElementsForWrite(elements)) return@run
              val progressTask = SequentialModalProgressTask(project, INFER_SUPPORT_ANNOTATIONS, false)
              progressTask.setMinIterationTime(200)
              progressTask.setTask(AnnotateTask(project, progressTask, infos))
              ProgressManager.getInstance().run(progressTask)
            } else {
              nothingFoundMessage(project)
            }
          }
        } finally {
          action.finish()
        }
      }
    }

    private fun showUsageView(project: Project, usageInfos: Array<UsageInfo>, scope: AnalysisScope) {
      val targets = UsageTarget.EMPTY_ARRAY
      val convertUsagesRef = Ref<Array<UsageInfo2UsageAdapter>>()
      if (!ProgressManager.getInstance().runProcessWithProgressSynchronously(
          { ApplicationManager.getApplication().runReadAction { convertUsagesRef.set(UsageInfo2UsageAdapter.convert(usageInfos)) } },
          "Preprocess usages", true, project
        )
      ) {
        return
      }
      if (convertUsagesRef.isNull) return
      val usages = convertUsagesRef.get()
      val presentation = UsageViewPresentation()
      presentation.tabText = "Infer Nullity Preview"
      presentation.isShowReadOnlyStatusAsRed = true
      presentation.isShowCancelButton = true
      presentation.searchString = RefactoringBundle.message("usageView.usagesText")
      val usageView = UsageViewManager.getInstance(project).showUsages(targets, usages, presentation, rerunFactory(project, scope))
      val refactoringRunnable = applyRunnable(project) {
        val infos = UsageViewUtil.getNotExcludedUsageInfos(usageView)
        infos.toTypedArray()
      }
      val canNotMakeString =
        "Cannot perform operation.\nThere were changes in code after usages have been found.\nPlease perform operation search again."
      usageView.addPerformOperationAction(
        refactoringRunnable,
        INFER_SUPPORT_ANNOTATIONS,
        canNotMakeString,
        INFER_SUPPORT_ANNOTATIONS,
        false
      )
    }

    private fun rerunFactory(project: Project, scope: AnalysisScope): Factory<UsageSearcher> {
      return Factory {
        object : UsageInfoSearcherAdapter() {
          override fun findUsages(): Array<UsageInfo> {
            return findUsages(project, scope, scope.fileCount) ?: UsageInfo.EMPTY_ARRAY
          }

          override fun generate(processor: Processor<in Usage>) {
            processUsages(processor, project)
          }
        }
      }
    }

    private fun addDependency(module: Module, libraryCoordinate: String) {
      if (StringUtil.isNotEmpty(libraryCoordinate)) {
        ModuleRootModificationUtil.updateModel(module) {
          val buildModel = GradleBuildModel.get(module)
          if (buildModel != null) {
            val name =
              GradleUtil.mapConfigurationName(CommonConfigurationNames.COMPILE, GradleUtil.getAndroidGradleModelVersionInUse(module), false)
            buildModel.dependencies().addArtifact(name, libraryCoordinate)
            buildModel.applyChanges()
          }
        }
      }
    }
  }
}